sssd_test_framework/utils/authentication.py

```diff
@@ -397,44 +407,82 @@ def passkey(self, username: str, *, pin: str | int, device: str, ioctl: str, scr
+                exit $code
+            }}
+
             # It takes some time to get authentication failure
             set timeout {DEFAULT_AUTHENTICATION_TIMEOUT}
             set prompt "\n.*\[#\$>\] $"
+            set command "{command}"
 
-            spawn "{playback_umockdev}"
+            spawn {playback_umockdev}
 
             expect {{
-                "Insert your passkey device, then press ENTER*" {{send -- "\r"}}
-                timeout {{puts "expect result: Unexpected output"; exit 201}}
-                eof {{puts "expect result: Unexpected end of file"; exit 202}}
+                "Insert your passkey device, then press ENTER*" {{send -- "\n"}}
+                timeout {{exitmsg "Unexpected output" 201}}
+                eof {{exitmsg "Unexpected end of file" 202}}
             }}
 
             expect {{
                 "Enter PIN:*" {{send -- "{pin}\r"}}
-                timeout {{puts "expect result: Unexpected output"; exit 201}}
-                eof {{puts "expect result: Unexpected end of file"; exit 202}}
+                timeout {{exitmsg "Unexpected output" 211}}
+                eof {{exitmsg "Unexpected end of file" 212}}
             }}
 
             expect {{
-                -re $prompt {{puts "expect result: Password authentication successful"; exit 0}}
-                "Authentication failure" {{puts "expect result: Authentication failure"; exit 1}}
-                timeout {{puts "expect result: Unexpected output"; exit 201}}
-                eof {{puts "expect result: Unexpected end of file"; exit 202}}
+                -re $prompt {{
+                    if {{ $command eq "" }} {{ exitmsg "Password authentication successful" 0}}
+                }}
+                "Authentication failure" {{exitmsg "Authentication failure" 1}}
+                timeout {{exitmsg "Unexpected output" 221}}
+                eof {{exitmsg "Unexpected end of file" 222}}
             }}
 
-            puts "expect result: Unexpected code path"
-            exit 203
-            """
-        )
+            send -- "$command\n"
+
+            expect {{
+                "{expected_output}" {{exitmsg "Got the TGT for user" 200}}
+                timeout {{exitmsg "Unexpected output" 231}}
+                eof {{exitmsg "Unexpected end of file" 232}}
+            }}
+
+            exitmsg "Unexpected code path" 203
+
+            """,
+            verbose=False,
+        ) 
 
         self.fs.restore("/usr/libexec/sssd/passkey_child")
 
         if result.rc > 200:
             raise ExpectScriptError(result.rc)
 
-        return result.rc == 0
+        expect_data = result.stdout_lines[-3:]
+
+        # Get command exit code.
+        cmdrc = int(expect_data[2].split(":")[1].strip())
+
+        # Alter stdout, first line is spawned command, the last three are our expect output.
+        stdout = "\n".join(result.stdout_lines[1:-3])
+
+        return result.rc, cmdrc, stdout, result.stderr
```

Comment:
by changing the return,  you'll need to make sure you change the code that uses this method.

Reply:
We need this to get console messages.